dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/PoolingLayer.java

```diff
@@ -0,0 +1,144 @@
+import org.apache.reef.tang.annotations.Parameter;
+
+import javax.inject.Inject;
+
+/**
+ * Pooling layer.
+ *
+ * This layer is not learnable.
+ * This layer resizes input matrix spatially, using max pooling or average pooling.
+ * In a forward pass,
+ * max pooling picks maximum value in certain range ( kernalHeight * kernalWidth) and these values make up output.
+ * average pooling get average of values in certain range (kernalHeight * kernalWidth) and these values make up output.
+ * In a backward pass,
+ * each value of error matrix is the sum of elements in next error matrix
+ * on which had an impact in feedforward step.
+ */
+
+public abstract class PoolingLayer extends LayerBase {
+
+  private final int[] outputShape;
+  private Matrix trackMatrix;
+  private Function poolingFunctions;
+
+  @Inject
+  public PoolingLayer(@Parameter(LayerConfigurationParameters.LayerIndex.class) final int index,
+                      @Parameter(LayerConfigurationParameters.LayerInputShape.class) final String inputShape,
+                      final LayerParameterInitializer layerParameterInitializer) {
+    super(index, inputShape);
+    this.outputShape = setOutputShape();
+    setLayerParameter(layerParameterInitializer.generateInitialParameter());
+  }
+
+  @Override
+  public int[] getOutputShape() {
+    return outputShape;
+  }
+
+  /**
+   * This function computes output size.
+   * input size: row * col
+   * output size: row' * col'
+   * row = (row − kernal_height) / stride + 1
+   * col = (col − kernal_width) / stride + 1
+   */
+
+  public int[] setOutputShape() {
+    final int[] inputShape = getInputShape();
+    int[] computedShape;
+    switch (inputShape.length) {
+    case 1:
+      computedShape = new int[1];
+      computedShape[0] = (inputShape[0] - getLayerParameter().getKernalHeight()) / getLayerParameter().getStride() + 1;
+      return computedShape;
+    case 2:
+      computedShape = new int[2];
+      computedShape[0] = (inputShape[0] - getLayerParameter().getKernalHeight()) / getLayerParameter().getStride() + 1;
+      computedShape[1] = (inputShape[1] - getLayerParameter().getKernalWidth()) / getLayerParameter().getStride() + 1;
+      return computedShape;
+    default:
+      throw new IllegalArgumentException("Unsupported input dimension: " + Integer.toString(inputShape.length));
+    }
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public boolean isLearnable() {
+    return false;
+  }
+
+  public abstract Matrix feedForwardMaxPooling(final Matrix input);
+
+  public abstract Matrix feedForwardAveragePooling(final Matrix input);
+
+  /**
+   * Computes output values for this pooling layer.
+   * available pooling type: max, average
+   * @param input input values for this layer.
+   * @return output values for this layer.
+   */
+
+  @Override
+  public  Matrix feedForward(final Matrix input) {
+    switch (getLayerParameter().getPoolingType().toLowerCase()) {
```

Comment:
Why don't we use `enum`? It will be more efficient than changing a string to lower case and comparing it with `"max"` and `"average"`.